File: engine/src/warden/engine.py
"""Engine — junta registry + adapters + event bus, mantém instância viva por projeto."""

import subprocess
from collections.abc import Callable
from dataclasses import dataclass
from pathlib import Path

from warden.adapters.base import Adapter, ProcessStatus
from warden.adapters.factory import create_adapter
from warden.bus import EventBus
from warden.config import ProjectConfig
from warden.events import Event, EventType
from warden.file_error_watcher import FileErrorWatcher
from warden.git import (
    CONFIRM_VERBS,
    GitCommandResult,
    GitInfo,
    git_command,
    git_info,
)
from warden.git_watcher import GitWatcher
from warden.languages import detect_languages
from warden.notifier import Notifier, NullNotifier
from warden.registry import Registry
from warden.store import EventStore


@dataclass
class ActionResult:
    exit_code: int
    output: str


class ConfirmationRequired(Exception):
    """Ação marcada `destructive = true` chamada sem `confirmed=True`."""


class Engine:
    def __init__(
        self,
        config_dir: Path,
        store: EventStore | None = None,
        notifier: Notifier | None = None,
    ):
        self.registry = Registry(config_dir)
        self.bus = EventBus()
        self.store = store
        self.notifier = notifier or NullNotifier()
        if store is not None:
            self.bus.subscribe(store.record)
        self.bus.subscribe(self._maybe_notify)
        self._adapters: dict[str, Adapter] = {}
        self._file_watchers: list[FileErrorWatcher] = []
        self._git_watchers: list[GitWatcher] = []

    def _maybe_notify(self, event: Event) -> None:
        try:
            project = self.registry.get(event.project_id)
        except KeyError:
            return
        if event.type == EventType.ERROR and project.notify.on_error:
            self.notifier.notify(event, project)
        elif event.type == EventType.FINISHED and project.notify.on_finished:
            self.notifier.notify(event, project)
        elif event.type == EventType.GIT_BEHIND and project.notify.on_git_behind:
            self.notifier.notify(event, project)

    def boot(self) -> None:
        self.registry.load()
        self._start_file_watchers()
        self._start_git_watchers()

    def shutdown(self) -> None:
        for watcher in self._file_watchers:
            watcher.stop()
        self._file_watchers.clear()
        for watcher in self._git_watchers:
            watcher.stop()
        self._git_watchers.clear()

    def _start_file_watchers(self) -> None:
        for project in self.registry.all():
            for log_source in project.log_sources:
                if log_source.type != "file" or not log_source.error_patterns:
                    continue
                watcher = FileErrorWatcher(
                    path=self._resolve_log_path(project, log_source.path),
                    patterns=log_source.error_patterns,
                    on_error=self._make_file_error_handler(project.id, log_source.name),
                )
                watcher.start()
                self._file_watchers.append(watcher)

    def _start_git_watchers(self) -> None:
        for project in self.registry.all():
            if not project.git.watch:
                continue
            watcher = GitWatcher(
                path=Path(project.path),
                remote=project.git.remote,
                interval=project.git.interval,
                on_behind=self._make_git_behind_handler(project.id, project.git.remote),
            )
            watcher.start()
            self._git_watchers.append(watcher)

    def _make_git_behind_handler(self, project_id: str, remote: str) -> Callable[[int], None]:
        def handler(behind: int) -> None:
            self.bus.publish(
                Event(
                    project_id=project_id,
                    type=EventType.GIT_BEHIND,
                    message=f"{behind} commit(s) atrás de {remote}",
                )
            )

        return handler

    @staticmethod
    def _resolve_log_path(project: ProjectConfig, log_path: str | None) -> Path:
        assert log_path is not None
        path = Path(log_path)
        return path if path.is_absolute() else Path(project.path) / path

    def _make_file_error_handler(
        self, project_id: str, source_name: str
    ) -> Callable[[str], None]:
        def handler(entry: str) -> None:
            self.bus.publish(
                Event(
                    project_id=project_id,
                    type=EventType.ERROR,
                    message=f"[{source_name}] {entry[:500]}",
                )
            )

        return handler

    def _adapter(self, project_id: str) -> Adapter:
        if project_id not in self._adapters:
            config = self.registry.get(project_id)
            adapter = create_adapter(config)
            adapter.set_on_exit(lambda returncode: self._handle_exit(project_id, returncode))
            self._adapters[project_id] = adapter
        return self._adapters[project_id]

    def _handle_exit(self, project_id: str, returncode: int) -> None:
        event_type = EventType.FINISHED if returncode == 0 else EventType.ERROR
        self.bus.publish(
            Event(project_id=project_id, type=event_type, message=f"exit={returncode}")
        )

    def start(self, project_id: str) -> None:
        self._adapter(project_id).start()
        self.bus.publish(Event(project_id=project_id, type=EventType.STARTED))

    def stop(self, project_id: str) -> None:
        self._adapter(project_id).stop()
        self.bus.publish(Event(project_id=project_id, type=EventType.STOPPED))

    def status(self, project_id: str) -> ProcessStatus:
        return self._adapter(project_id).status()

    def logs(self, project_id: str, tail: int = 100, service: str | None = None) -> list[str]:
        return self._adapter(project_id).logs(tail, service=service)

    def services(self, project_id: str) -> list[str]:
        return self._adapter(project_id).services()

    def git_info(self, project_id: str) -> GitInfo | None:
        project = self.registry.get(project_id)
        return git_info(Path(project.path))

    def git_command(
        self, project_id: str, verb: str, confirmed: bool = False
    ) -> GitCommandResult:
        project = self.registry.get(project_id)
        if verb in CONFIRM_VERBS and not confirmed:
            raise ConfirmationRequired(f"verbo git {verb!r} exige confirmação")
        return git_command(Path(project.path), verb)

    def languages(self, project_id: str) -> list[str]:
        project = self.registry.get(project_id)
        return detect_languages(Path(project.path))

    def history(self, project_id: str, limit: int = 50) -> list[dict]:
        if self.store is None:
            return []
        return self.store.history(project_id, limit)

    def run_action(
        self, project_id: str, action_name: str, confirmed: bool = False
    ) -> ActionResult:
        project = self.registry.get(project_id)
        try:
            action = next(a for a in project.actions if a.name == action_name)
        except StopIteration:
            raise KeyError(f"ação {action_name!r} não encontrada em {project_id!r}") from None
        if action.interactive:
            raise ValueError(f"ação {action_name!r} é interativa, não suportada via API")
        if action.destructive and not confirmed:
            raise ConfirmationRequired(
                f"ação {action_name!r} é destrutiva — chame com confirmed=True"
            )
        result = subprocess.run(
            action.cmd, cwd=project.path, capture_output=True, text=True, timeout=300
        )
        if action.destructive and self.store is not None:
            self.store.record_action(
                project_id, action_name, action.cmd, confirmed, result.returncode
            )
        return ActionResult(exit_code=result.returncode, output=result.stdout + result.stderr)

    def action_audit(self, project_id: str, limit: int = 50) -> list[dict]:
        if self.store is None:
            return []
        return self.store.action_audit(project_id, limit)

File: agent/Warden.Domain/Engine.cs
using System.Threading.Channels;
using Warden.Domain.Adapters;
using Warden.Domain.Config;
using Warden.Domain.Discovery;
using Warden.Domain.Events;
using Warden.Domain.Languages;
using Warden.Domain.Notify;
using Warden.Domain.Trust;
using Warden.Domain.Vitals;
using Warden.Domain.Watch;

namespace Warden.Domain;

/// <summary>
/// Facade que coordena Registry + AdapterPool + ManifestRegistry + WatcherCoordinator + EventBus.
/// Cada responsabilidade interna vive na sua própria classe; o Engine só orquestra o fluxo de alto
/// nível (equivalente ao `Engine` do engine Python — ver NEW_CONTEXT.md §12 fase 8).
/// </summary>
public sealed class Engine
{
    private readonly Registry _registry;
    private readonly ManifestRegistry _manifestRegistry;
    private readonly IEventStore? _store;
    private readonly INotifier _notifier;
    private readonly AdapterPool _adapterPool;
    private readonly WatcherCoordinator _watchers;
    private readonly SystemVitalsSampler _systemVitalsSampler = new();
    private ProjectsWatcher? _projectsWatcher;

    public EventBus Bus { get; } = new();

    public Engine(Registry registry, ManifestRegistry manifestRegistry, IEventStore? store = null, INotifier? notifier = null)
    {
        _registry = registry;
        _manifestRegistry = manifestRegistry;
        _store = store;
        _notifier = notifier ?? new NullNotifier();

        _adapterPool = new AdapterPool(_registry);
        _adapterPool.SetOnExit((projectId, returnCode) =>
        {
            var type = returnCode == 0 ? EventType.Finished : EventType.Error;
            Bus.Publish(new Event(projectId, type, $"exit={returnCode}"));
        });

        _watchers = new WatcherCoordinator(Bus.Publish,
            (projectId, actionName) => RunAction(projectId, actionName, confirmed: false));

        if (_store is not null) Bus.Subscribe(_store.Record);
        Bus.Subscribe(MaybeNotify);
    }

    public void Boot(TimeSpan? projectsWatchInterval = null)
    {
        _registry.Load();
        foreach (var project in _registry.All())
            _manifestRegistry.Refresh(project);

        _watchers.StartAll(_registry.All());
        _projectsWatcher = new ProjectsWatcher(_registry.ProjectsDir, ReloadRegistry, projectsWatchInterval);
        _projectsWatcher.Start();
        _systemVitalsSampler.Prime();
    }

    /// <summary>
    /// Shutdown gracioso do daemon. <see cref="ReloadRegistry"/> chama <see cref="WatcherCoordinator.Restart"/>
    /// diretamente — não este método — para evitar que o callback do ProjectsWatcher tente dar Join
    /// nele mesmo.
    /// </summary>
    public void Shutdown()
    {
        _projectsWatcher?.Stop();
        _projectsWatcher = null;
        _watchers.StopAll();
    }

    public IReadOnlyList<ProjectConfig> AllProjects() => _registry.All();

    public ProjectConfig GetProject(string projectId) => _registry.Get(projectId);

    /// <summary>Start passa pelo trust gate (§10.3/10.4) — o comando [start] faz parte do manifesto.</summary>
    public void Start(string projectId)
    {
        if (Manifest(projectId).Status != TrustStatus.Approved)
            throw new ManifestNotApprovedException(
                $"projeto \"{projectId}\" não está aprovado — aprove no Admin antes de iniciar");

        _adapterPool.Get(projectId).Start();
        Bus.Publish(new Event(projectId, EventType.Started));
    }

    public void Stop(string projectId)
    {
        _adapterPool.Get(projectId).Stop();
        Bus.Publish(new Event(projectId, EventType.Stopped));
    }

    public ProcessStatus Status(string projectId) => _adapterPool.Get(projectId).Status();

    public IReadOnlyList<string> Logs(string projectId, int tail = 100, string? service = null) =>
        _adapterPool.Get(projectId).Logs(tail, service);

    /// <summary>
    /// Retorna um <see cref="ChannelReader{T}"/> que emite linhas em tempo real conforme chegam do
    /// processo. Null quando o adapter não suporta streaming (ex: Docker) — o Hub deve fazer fallback
    /// para polling via <see cref="Logs"/>. O reader é fechado quando o <paramref name="cancellation"/>
    /// é cancelado (desconexão do Hub) ou quando o processo termina.
    /// </summary>
    public ChannelReader<string>? SubscribeLogs(string projectId, CancellationToken cancellation) =>
        _adapterPool.Get(projectId).LogBroadcaster?.Subscribe(cancellation);

    public IReadOnlyList<string> Services(string projectId) => _adapterPool.Get(projectId).Services();

    public ProjectManifestSnapshot Manifest(string projectId) =>
        _manifestRegistry.Get(projectId)
        ?? throw new KeyNotFoundException($"manifesto do projeto \"{projectId}\" ainda não foi resolvido");

    /// <summary>Aprovação só por caminho local/confiável (Admin). Sem rota HTTP no Console — ver NEW_CONTEXT.md §10.3.</summary>
    public ProjectManifestSnapshot Approve(string projectId) => _manifestRegistry.Approve(GetProject(projectId));

    public ActionExecutionResult RunAction(string projectId, string actionName, bool confirmed)
    {
        var snapshot = Manifest(projectId);
        var command = snapshot.Manifest.Commands.FirstOrDefault(c => c.Name == actionName && c.Name != "start")
            ?? throw new KeyNotFoundException($"ação \"{actionName}\" não encontrada em \"{projectId}\"");

        if (snapshot.Status != TrustStatus.Approved)
            throw new ManifestNotApprovedException(
                $"projeto \"{projectId}\" está com status {snapshot.Status} — aprove no Admin antes de executar ações");

        if (command.Interactive)
            throw new ActionInteractiveException($"ação \"{actionName}\" é interativa, não suportada via API");

        if (command.Destructive && !confirmed)
            throw new ConfirmationRequiredException($"ação \"{actionName}\" é destrutiva — exige confirm=true");

        var result = CommandExecutor.Run(command);
        if (command.Destructive)
            _store?.RecordAction(projectId, actionName, command.Argv, confirmed, result.ExitCode);

        return result;
    }

    public SystemVitalsInfo SystemVitals() => _systemVitalsSampler.Sample();

    public IReadOnlyList<string> Languages(string projectId) =>
        LanguageDetector.Detect(GetProject(projectId).Path);

    public Git.GitInfo? GitInfo(string projectId) => Git.GitService.Info(GetProject(projectId).Path);

    /// <summary>Verbos mutantes (pull/push) exigem confirmed=true — mesma semântica 409 das ações destrutivas.</summary>
    public Git.GitCommandResult GitCommand(string projectId, string verb, bool confirmed = false)
    {
        var project = GetProject(projectId);
        if (Git.GitService.ConfirmVerbs.Contains(verb) && !confirmed)
            throw new ConfirmationRequiredException($"verbo git \"{verb}\" exige confirmação");

        return Git.GitService.Command(project.Path, verb, project.Git.Remote);
    }

    public IReadOnlyList<HistoryEntry> History(string projectId, int limit = 50) =>
        _store?.History(projectId, limit) ?? [];

    public IReadOnlyList<ActionAuditEntry> ActionAudit(string projectId, int limit = 50) =>
        _store?.ActionAudit(projectId, limit) ?? [];

    // ---- Discovery / Scaffold ----

    private string GlobalConfigPath => Path.Combine(_registry.ConfigDir, "config.toml");

    private string ProjectTomlPath(string projectId) =>
        Path.Combine(_registry.ConfigDir, "projects", $"{projectId}.toml");

    public GlobalConfig LoadGlobalConfig() => ConfigLoader.LoadGlobalConfig(GlobalConfigPath);

    public GlobalConfig AddScanPath(string path) => ProjectDiscovery.AddScanPath(_registry.ConfigDir, path);

    public GlobalConfig RemoveScanPath(string path) => ProjectDiscovery.RemoveScanPath(_registry.ConfigDir, path);

    public IReadOnlyList<DiscoveredProject> Discover() => ProjectDiscovery.DiscoverProjects(LoadGlobalConfig(), _registry);

    public BrowseResult Browse(string? path) => ProjectDiscovery.BrowseDirectory(path);

    public ProjectConfig PreviewConfig(string path, string? projectId) => Scaffold.BuildConfig(path, projectId);

    public ProjectConfig GetProjectConfigFile(string projectId)
    {
        var target = ProjectTomlPath(projectId);
        if (!File.Exists(target))
            throw new KeyNotFoundException($"config de \"{projectId}\" não encontrada");

        return ConfigLoader.LoadProjectConfig(target);
    }

    public string ApplyConfig(ProjectConfig config)
    {
        var toml = Scaffold.RenderToml(config);
        var target = ProjectTomlPath(config.Id);
        Directory.CreateDirectory(Path.GetDirectoryName(target)!);
        File.WriteAllText(target, toml);
        ReloadRegistry();
        return toml;
    }

    /// <summary>
    /// Recarrega config do disco sem restart do daemon. Preserva adapters de processos rodando para não
    /// perder referência ao processo em execução.
    /// </summary>
    public void ReloadRegistry()
    {
        _registry.Load();
        foreach (var project in _registry.All())
            _manifestRegistry.Refresh(project);

        _adapterPool.PruneStopped();
        _watchers.Restart(_registry.All());
    }

    private void MaybeNotify(Event @event)
    {
        ProjectConfig project;
        try { project = _registry.Get(@event.ProjectId); }
        catch (KeyNotFoundException) { return; }

        var shouldNotify = @event.Type switch
        {
            EventType.Error => project.Notify.OnError,
            EventType.Finished => project.Notify.OnFinished,
            EventType.GitBehind => project.Notify.OnGitBehind,
            _ => false,
        };
        if (shouldNotify) _notifier.Notify(@event, project);
    }
}
